### PR TITLE
[MKL-DNN] Thread-Safety for MKL-DNN reusing Part 1

### PR DESCRIPTION
**Pull request by jczaja**
This PR adress the issue reported in #17611 by making Reusing concept thread safe. Currently only
part of ops were adapted. After sorting out an issue #17960  additional PR for Relu, pooling and INT8 prims will be raised.

@luotao1 I have updated this branch couple of hours ago, so please double check if this PR works for you. If ContentDNN is not working then it means that Relu and other missing ops has to be updated as well. Anyway please share error messages if any.

### CODE REVIEW

**Comment by ghost** [comment]
@LeoZhao-Intel Please help to review this PR. Thanks!

**Comment by luotao1** [comment]
I test it on face deployment service, it works successfully.

File: paddle/fluid/platform/mkldnn_reuse.h [hunk 1]
```diff
@@ -696,35 +663,43 @@ class ConvMKLDNNTemplateHandler : public MKLDNNHandler {
       const bool fuse_relu, const bool fuse_residual_conn,
       const bool fuse_brelu, const float fuse_brelu_threshold,
       mkldnn::prop_kind fwd_prop_kind) {
-    const std::string key_conv_pd = key_ + "@conv_pd";
+    // Conv PD has to be passed to Grad op that
+    // may be exxecuted by diffrent thread, hence
```
Comment (LeoZhao-BJ) [inline paddle/fluid/platform/mkldnn_reuse.h:667]: typo, "executed"

**Review by LeoZhao-BJ — COMMENTED** [review]
Basically LGTM.

Few comments for Part2:
1. By key generation with threadid, is lock still needed?
2. Current softmax grad op is dependent on softmax in mkldnn implementation, unlike transpose2 does, can we remove this tight pairing?

**Review by luotao1 — APPROVED** [review]
LGTM! Please refine https://github.com/PaddlePaddle/Paddle/pull/17965#pullrequestreview-247957086 in next PR!

**Comment by luotao1** [comment]
@jczaja @jianhang-liu @LeoZhao-Intel I test this PR on content DNN. It runs successfully on multi-thread with transpose_mkldnn_op.

**Comment by LeoZhao-BJ** [comment]
I did similar test before, both transpose2/transpose2_grad on mkldnn can work well in ParallelExecutor.

**Comment by jczaja** [comment]
@luotao1 Thanks for letting us know on your findings

**Pull request merged**